File: backend/app/models/candle.py
from typing import Optional

from pydantic import BaseModel


class Candle(BaseModel):
  """A single OHLCV candle. Timestamp is Unix milliseconds."""

  timestamp: int
  open: float
  high: float
  low: float
  close: float
  volume: float


class ChartMarker(BaseModel):
  """An entry/exit marker derived from a trade activity."""

  timestamp: int
  type: str
  side: str
  price: float
  quantity: float


class ChartLevels(BaseModel):
  """Price levels overlaid on the chart."""

  stop_loss: Optional[float] = None
  take_profits: list[float] = []


class ChartWindow(BaseModel):
  """Inclusive date range (ISO YYYY-MM-DD) covered by the candle query."""

  start: str
  end: str


class TradeChartData(BaseModel):
  """Chart payload for a single trade: candles, markers, levels and window."""

  ticker: str
  resolution: str
  candles: list[Candle]
  markers: list[ChartMarker]
  levels: ChartLevels
  window: ChartWindow


class CandleMeta(BaseModel):
  """Metadata explaining why `data` is null or candles are empty.

  reason is one of:
  - 'no_ticker': the trade's asset has no massive_ticker (data is null)
  - 'pending':   no candles are available yet for the range (candles empty)
  """

  reason: str


class TradeCandlesResponse(BaseModel):
  """Envelope for GET /api/trades/{id}/candles."""

  data: Optional[TradeChartData] = None
  meta: Optional[CandleMeta] = None
  error: Optional[str] = None

File: frontend/src/types/candle.ts
/**
 * Types for the trade chart / candle endpoint.
 *
 * Field names mirror the API's snake_case JSON so responses map directly
 * with no transform layer.
 */

export interface Candle {
  timestamp: number
  open: number
  high: number
  low: number
  close: number
  volume: number
}

export interface ChartMarker {
  timestamp: number
  type: 'entry' | 'exit'
  side: 'Buy' | 'Sell'
  price: number
  quantity: number
}

export interface ChartLevels {
  stop_loss: number | null
  take_profits: number[]
}

export interface ChartWindow {
  start: string
  end: string
}

export interface TradeChartData {
  ticker: string
  resolution: string
  candles: Candle[]
  markers: ChartMarker[]
  levels: ChartLevels
  window: ChartWindow
}

export interface CandleMeta {
  reason: string
}

export interface TradeCandlesResponse {
  data: TradeChartData | null
  meta: CandleMeta | null
  error: string | null
}

export type ChartResolution = 'M1' | 'M5' | 'M15' | 'H1' | 'H4' | 'D1'

export const CHART_RESOLUTIONS: ChartResolution[] = ['M1', 'M5', 'M15', 'H1', 'H4', 'D1']
